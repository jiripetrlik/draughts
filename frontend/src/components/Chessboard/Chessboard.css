table.chessboard {
    border-collapse: collapse;
}
