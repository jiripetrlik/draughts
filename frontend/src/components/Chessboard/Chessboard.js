import React from "react"
import Chessboardsquare from "../ChessboardSquare/ChessboardSquare"
import "./Chessboard.css"

class Chessboard extends React.Component {

    constructor(props) {
        super(props)
        this.size = parseInt(props.size)
        this.state = {
            selected: null,
            possibleMoves: []
        }

        this.handleSquareClick = this.handleSquareClick.bind(this)
    }

    createRow(index) {
        let cells = []
        for (var i = 0; i < this.size; i++) {
            cells.push(<Chessboardsquare key={i} row={index} column={i} pieceType={this.pieceType(index, i)}
                selected={this.isSelected(index, i)} onHandleClick={this.handleSquareClick}/>)
        }

        return (
            <tr key={index}>
                <td>{index + 1}</td>
                {cells}
            </tr>
        )
    }
    
    pieceType(row, column) {
        for (const i in this.props.pieces.Whitepieces) {
            const coordinates = this.props.pieces.Whitepieces[i]

            if (coordinates.X === row && coordinates.Y === column) {
                return "white-piece"
            }
        }

        for (const i in this.props.pieces.Blackpieces) {
            const coordinates = this.props.pieces.Blackpieces[i]

            if (coordinates.X === row && coordinates.Y === column) {
                return "black-piece"
            }
        }

        for (const i in this.props.pieces.Whitequeens) {
            const coordinates = this.props.pieces.Whitequeens[i]

            if (coordinates.X === row && coordinates.Y === column) {
                return "white-queen"
            }
        }

        for (const i in this.props.pieces.Blackqueens) {
            const coordinates = this.props.pieces.Blackqueens[i]

            if (coordinates.X === row && coordinates.Y === column) {
                return "black-queen"
            }
        }

        return "none"
    }

    pieceTypeOrMove(row, column) {
        let type = this.pieceType(row, column)
        
        for (const i in this.state.possibleMoves) {
            const coordinates = this.state.possibleMoves[i]

            if (coordinates.X === row && coordinates.Y === column) {
                return "move"
            }
        }

        return type
    }

    isSelected(x, y) {
        console.log("Is selected called for: " + x + " " + y)
        if (this.state.selected != null) {
            if ((x === this.state.selected.X) && (y === this.state.selected.Y)) {
                return "yes"
            }
        }

        return "no"
    }

    containsPiece(x, y) {
        let p
        if (this.props.player === "white") {
            p = this.props.pieces.Whitepieces
        } else {
            p = this.props.pieces.Blackpieces
        }

        for (const i in p) {
            const coordinates = p[i]
            if ((coordinates.X === x) && (coordinates.Y === y)) {
                return true
            }
        }

        return false
    }
    
    handleSquareClick(x, y) {
        if (this.containsPiece(x, y)) {
            this.setState({
                selected: {
                    X: x,
                    Y: y,
                }
            })
        } else {
            this.setState({ selected: null})
        }

        console.log(this.containsPiece(x, y))
    }

    render() {
        let rows = []
        if (this.props.player === "white") {
            for (let i = this.size - 1; i >= 0; i--) {
                rows.push(this.createRow(i))
            }
        } else {
            for (let i = 0; i < this.size; i++) {
                rows.push(this.createRow(i))
            }
        }

        let indexes = []
        indexes.push(<td key="1"></td>)
        for (let i = 65; i < 73; i++) {
            indexes.push(<td key={i}>{String.fromCharCode(i)}</td>)
        }

        return (
            <table className="chessboard">
                <tbody>
                    {rows}
                    <tr>{indexes}</tr>
                </tbody>
            </table>
        )
    }
}

export default Chessboard;
