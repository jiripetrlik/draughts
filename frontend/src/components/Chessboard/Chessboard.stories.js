import React from 'react'
import Chessboard from "./Chessboard"

export default {
  component: Chessboard,
  title: 'Chessboard',
};

const Template = args => <Chessboard {...args} />;

export const Default = Template.bind({});
Default.args = {
  size: 8,
  player: "white",
  pieces: {
    Whitepieces: [{X: 0, Y: 0}, {X: 0, Y: 2}],
    Blackpieces: [{X: 6, Y: 2}, {X: 5, Y: 3}, {X: 4, Y: 4}],
    Whitequeens: [{X: 4, Y: 6}],
    Blackqueens: [{X: 6, Y: 6}, {X: 4, Y: 4}],
  },
}

export const Empty = Template.bind({});
Empty.args = {
  size: 8,
  player: "white",
  pieces: {
    Whitepieces: [],
    Blackpieces: [],
    Whitequeens: [],
    Blackqueens: [],
  },
}

export const WhitePlayer = Template.bind({});
WhitePlayer.args = {
  size: 8,
  player: "white",
  pieces: {
    Whitepieces: [{X: 0, Y: 0}, {X: 0, Y: 2}, {X: 0, Y: 4}, {X: 0, Y: 6}, {X: 1, Y: 1}, {X: 1, Y: 3}, {X: 1, Y: 5}, {X: 1, Y: 7}],
    Blackpieces: [{X: 6, Y: 0}, {X: 6, Y: 2}, {X: 6, Y: 4}, {X: 6, Y: 6}, {X: 7, Y: 1}, {X: 7, Y: 3}, {X: 7, Y: 5}, {X: 7, Y: 7}],
    Whitequeens: [],
    Blackqueens: [],
  },
}

export const BlackPlayer = Template.bind({});
BlackPlayer.args = {
  size: 8,
  player: "black",
  pieces: {
    Whitepieces: [{X: 0, Y: 0}, {X: 0, Y: 2}, {X: 0, Y: 4}, {X: 0, Y: 6}, {X: 1, Y: 1}, {X: 1, Y: 3}, {X: 1, Y: 5}, {X: 1, Y: 7}],
    Blackpieces: [{X: 6, Y: 0}, {X: 6, Y: 2}, {X: 6, Y: 4}, {X: 6, Y: 6}, {X: 7, Y: 1}, {X: 7, Y: 3}, {X: 7, Y: 5}, {X: 7, Y: 7}],
    Whitequeens: [],
    Blackqueens: [],
  },
}
