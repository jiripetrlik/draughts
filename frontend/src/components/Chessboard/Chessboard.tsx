import React from "react"
import Chessboardsquare from "../ChessboardSquare/ChessboardSquare"
import "./Chessboard.css"
import {containsPieceWithColor, pieceType, possibleMoves} from "../../logic/index"

class Chessboard extends React.Component<Readonly<any>, Readonly<any>> {

    size: number

    constructor(props: any) {
        super(props)
        this.size = parseInt(props.size)
        this.state = {
            selected: null,
            possibleMoves: []
        }

        this.handleSquareClick = this.handleSquareClick.bind(this)
    }

    createRow(index: number) {
        let cells = []
        for (var i = 0; i < this.size; i++) {
            cells.push(<Chessboardsquare key={i} row={index} column={i}
                pieceType={this.pieceTypeOrMove(index, i)}
                selected={this.isSelected(index, i)} onHandleClick={this.handleSquareClick}/>)
        }

        return (
            <tr key={index}>
                <td>{index + 1}</td>
                {cells}
            </tr>
        )
    }

    pieceTypeOrMove(row: number, column: number) {
        let type = pieceType(row, column, this.props.pieces)
        
        for (const i in this.state.possibleMoves) {
            const coordinates = this.state.possibleMoves[i]

            if (coordinates.X === row && coordinates.Y === column) {
                return "move"
            }
        }

        return type
    }

    isSelected(x: number, y: number) {
        if (this.state.selected != null) {
            if ((x === this.state.selected.X) && (y === this.state.selected.Y)) {
                return "yes"
            }
        }

        return "no"
    }
    
    handleSquareClick(x: number, y: number) {
        if (containsPieceWithColor(x, y, this.props.player, this.props.pieces)) {
            this.setState({
                selected: {
                    X: x,
                    Y: y,
                },
                possibleMoves: possibleMoves(x, y, this.props.pieces, 8).map(m => m.destination)
            })
        } else {
            this.setState({ selected: null, possibleMoves: []})
        }
    }

    render() {
        let rows = []
        if (this.props.player === "white") {
            for (let i = this.size - 1; i >= 0; i--) {
                rows.push(this.createRow(i))
            }
        } else {
            for (let i = 0; i < this.size; i++) {
                rows.push(this.createRow(i))
            }
        }

        let indexes = []
        indexes.push(<td key="1"></td>)
        for (let i = 65; i < 73; i++) {
            indexes.push(<td key={i}>{String.fromCharCode(i)}</td>)
        }

        return (
            <table className="chessboard">
                <tbody>
                    {rows}
                    <tr>{indexes}</tr>
                </tbody>
            </table>
        )
    }
}

export default Chessboard;
