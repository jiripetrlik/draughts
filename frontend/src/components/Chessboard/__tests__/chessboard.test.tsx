import React from 'react';
import renderer from 'react-test-renderer';
import Chessboard from '../Chessboard'

test('snapshot test', () => {
    const size = 8
    const player = "white"
    const pieces = {
        Whitepieces: [{X: 0, Y: 0}, {X: 0, Y: 2}],
        Blackpieces: [{X: 6, Y: 2}, {X: 5, Y: 3}],
        Whitequeens: [{X: 4, Y: 6}],
        Blackqueens: [{X: 6, Y: 6}, {X: 4, Y: 4}],
    }

    const tree = renderer
        .create(<Chessboard size={size} player={player} pieces={pieces} />)
        .toJSON();
    expect(tree).toMatchSnapshot();
});
