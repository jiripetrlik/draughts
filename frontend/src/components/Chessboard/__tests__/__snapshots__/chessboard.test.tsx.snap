// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
<table
  className="chessboard"
>
  <tbody>
    <tr>
      <td>
        8
      </td>
      <td
        className="dark"
        onClick={[Function]}
      />
      <td
        className="light"
        onClick={[Function]}
      />
      <td
        className="dark"
        onClick={[Function]}
      />
      <td
        className="light"
        onClick={[Function]}
      />
      <td
        className="dark"
        onClick={[Function]}
      />
      <td
        className="light"
        onClick={[Function]}
      />
      <td
        className="dark"
        onClick={[Function]}
      />
      <td
        className="light"
        onClick={[Function]}
      />
    </tr>
    <tr>
      <td>
        7
      </td>
      <td
        className="light"
        onClick={[Function]}
      />
      <td
        className="dark"
        onClick={[Function]}
      />
      <td
        className="light"
        onClick={[Function]}
      >
        <img
          alt=""
          src="black-piece.svg"
        />
      </td>
      <td
        className="dark"
        onClick={[Function]}
      />
      <td
        className="light"
        onClick={[Function]}
      />
      <td
        className="dark"
        onClick={[Function]}
      />
      <td
        className="light"
        onClick={[Function]}
      >
        <img
          alt=""
          src="black-queen.svg"
        />
      </td>
      <td
        className="dark"
        onClick={[Function]}
      />
    </tr>
    <tr>
      <td>
        6
      </td>
      <td
        className="dark"
        onClick={[Function]}
      />
      <td
        className="light"
        onClick={[Function]}
      />
      <td
        className="dark"
        onClick={[Function]}
      />
      <td
        className="light"
        onClick={[Function]}
      >
        <img
          alt=""
          src="black-piece.svg"
        />
      </td>
      <td
        className="dark"
        onClick={[Function]}
      />
      <td
        className="light"
        onClick={[Function]}
      />
      <td
        className="dark"
        onClick={[Function]}
      />
      <td
        className="light"
        onClick={[Function]}
      />
    </tr>
    <tr>
      <td>
        5
      </td>
      <td
        className="light"
        onClick={[Function]}
      />
      <td
        className="dark"
        onClick={[Function]}
      />
      <td
        className="light"
        onClick={[Function]}
      />
      <td
        className="dark"
        onClick={[Function]}
      />
      <td
        className="light"
        onClick={[Function]}
      >
        <img
          alt=""
          src="black-queen.svg"
        />
      </td>
      <td
        className="dark"
        onClick={[Function]}
      />
      <td
        className="light"
        onClick={[Function]}
      >
        <img
          alt=""
          src="white-queen.svg"
        />
      </td>
      <td
        className="dark"
        onClick={[Function]}
      />
    </tr>
    <tr>
      <td>
        4
      </td>
      <td
        className="dark"
        onClick={[Function]}
      />
      <td
        className="light"
        onClick={[Function]}
      />
      <td
        className="dark"
        onClick={[Function]}
      />
      <td
        className="light"
        onClick={[Function]}
      />
      <td
        className="dark"
        onClick={[Function]}
      />
      <td
        className="light"
        onClick={[Function]}
      />
      <td
        className="dark"
        onClick={[Function]}
      />
      <td
        className="light"
        onClick={[Function]}
      />
    </tr>
    <tr>
      <td>
        3
      </td>
      <td
        className="light"
        onClick={[Function]}
      />
      <td
        className="dark"
        onClick={[Function]}
      />
      <td
        className="light"
        onClick={[Function]}
      />
      <td
        className="dark"
        onClick={[Function]}
      />
      <td
        className="light"
        onClick={[Function]}
      />
      <td
        className="dark"
        onClick={[Function]}
      />
      <td
        className="light"
        onClick={[Function]}
      />
      <td
        className="dark"
        onClick={[Function]}
      />
    </tr>
    <tr>
      <td>
        2
      </td>
      <td
        className="dark"
        onClick={[Function]}
      />
      <td
        className="light"
        onClick={[Function]}
      />
      <td
        className="dark"
        onClick={[Function]}
      />
      <td
        className="light"
        onClick={[Function]}
      />
      <td
        className="dark"
        onClick={[Function]}
      />
      <td
        className="light"
        onClick={[Function]}
      />
      <td
        className="dark"
        onClick={[Function]}
      />
      <td
        className="light"
        onClick={[Function]}
      />
    </tr>
    <tr>
      <td>
        1
      </td>
      <td
        className="light"
        onClick={[Function]}
      >
        <img
          alt=""
          src="white-piece.svg"
        />
      </td>
      <td
        className="dark"
        onClick={[Function]}
      />
      <td
        className="light"
        onClick={[Function]}
      >
        <img
          alt=""
          src="white-piece.svg"
        />
      </td>
      <td
        className="dark"
        onClick={[Function]}
      />
      <td
        className="light"
        onClick={[Function]}
      />
      <td
        className="dark"
        onClick={[Function]}
      />
      <td
        className="light"
        onClick={[Function]}
      />
      <td
        className="dark"
        onClick={[Function]}
      />
    </tr>
    <tr>
      <td />
      <td>
        A
      </td>
      <td>
        B
      </td>
      <td>
        C
      </td>
      <td>
        D
      </td>
      <td>
        E
      </td>
      <td>
        F
      </td>
      <td>
        G
      </td>
      <td>
        H
      </td>
    </tr>
  </tbody>
</table>
`;
